9.2.26 Lab: Dive into the uhashring repo

In [10]:
from uhashring import HashRing

# create a consistent hash ring of 3 nodes of weight 1
hr = HashRing(nodes=['node1', 'node2', 'node3'])

# get the node name for the 'coconut' key
target_node = hr.get_node('coconut')

print(target_node)

node2


In [12]:
# get the 'coconut' key from the right host's redis instance
hr.get('coconut')

{'hostname': 'node2',
 'instance': None,
 'nodename': 'node2',
 'port': None,
 'vnodes': 160,
 'weight': 1}

Where is the hash created? (in the ring.py file)

In [13]:
hr.get_node('coconut')

'node2'

In [17]:
hr.get_server('coconut')

(157285054189446768912416291781797853523, 'node2')

How many virtual nodes are created for each physical node?

In [18]:
hr.get('coconut')

{'hostname': 'node2',
 'instance': None,
 'nodename': 'node2',
 'port': None,
 'vnodes': 160,
 'weight': 1}

160 (see ring.py line 30)

Where are virtual nodes created?

What is the naming convention for virtual nodes?

In [19]:
hr.get_nodes()

dict_keys(['node1', 'node2', 'node3'])

Return all the ring positions for node 1.

In [22]:
hr.get_node_pos('node2')

216

Revisit the example from Basic usage and look at the target_n for the key 'coconut'.